File: node/src/nautilus.ts
import { Scpi, fromUri } from './scpi.js'
import { SerialPort } from 'serialport';

class Nautilus {

    scpi: Scpi;
    private serialBuffer: Buffer;

    constructor (uri: string = "tcp://nautilus.local") {
        this.scpi = fromUri(uri);
        this.serialBuffer = Buffer.alloc(0);
    }

    async open() {
        await this.scpi.open();

        try {
            await this.isPresent(true)
        }
        catch (err) {
            // If we get an error during open, we really want to leave the transport closed.
            await this.scpi.close()
            throw err;
        }
    }

    async close() {
        await this.reset()
        await this.scpi.close()
    }

    async isPresent(throwOnError: boolean = false): Promise<boolean> {
        const success = (await this.scpi.getIdn()).startsWith("TL Embedded, Nautilus,")
        if (throwOnError && !success) {
            throw new Error("Nautilus not found")
        }
        return success;
    }

    async reset() {
        await this.scpi.write("*RST");
    }

    // ADC input commands

    async getInputVoltage(channel: number): Promise<number> {
        assertChannel(channel, 12);
        return parseFloat( await this.scpi.query(`INP${channel}:VOLT?`) )
    }

    // DAC output commands

    async setOutputVoltage(channel: number, volts: number) {
        assertChannel(channel, 12)
        await this.scpi.write(`OUT${channel}:VOLT ${volts.toFixed(3)}V`)
    }

    async setOutputEnable(channel: number, enable: boolean) {
        assertChannel(channel, 12)
        await this.scpi.write(`OUT${channel}:ENA ${enable?"ON":"OFF"}`)
    }

    async setOutput(channel: number, enable: boolean, volts: number = 0.0) {
        assertChannel(channel, 12)
        await this.scpi.write(`OUT${channel}:SET ${enable?"ON":"OFF"}, ${volts.toFixed(3)}V`)
    }

    // PSU output commands

    async setPsuVoltage(channel: number, volts: number) {
        assertChannel(channel, 2)
        await this.scpi.write(`POW${channel}:VOLT ${volts.toFixed(3)}V`)
    }

    async setPsuCurrent(channel: number, amps: number) {
        assertChannel(channel, 2)
        await this.scpi.write(`POW${channel}:CURR ${amps.toFixed(3)}A`)
    }

    async setPsuEnable(channel: number, enable: boolean) {
        assertChannel(channel, 2)
        await this.scpi.write(`POW${channel}:ENA ${enable?"ON":"OFF"}`)
    }

    async setPsu(channel: number, enable: boolean, volts: number, amps: number) {
        //assertChannel(channel, 2)
        //await this.scpi.write(`POW${channel}:SET ${b2s(enable)}, ${n2s(volts)}V, ${n2s(amps)}A`)
        await this.setPsuVoltage(channel, volts)
        await this.setPsuCurrent(channel, amps)
        await this.setPsuEnable(channel, enable)
    }

    async getPsuVoltage(channel: number): Promise<number> {
        assertChannel(channel, 2)
        return parseFloat(await this.scpi.query(`POW${channel}:MEAS:VOLT?`))
    }

    async getPsuCurrent(channel: number): Promise<number> {
        assertChannel(channel, 2)
        return parseFloat(await this.scpi.query(`POW${channel}:MEAS:CURR?`))
    }

    // System commands

    async getVin(): Promise<number> {
        return parseFloat(await this.scpi.query("VIN?"))
    }

    async getTemperature(): Promise<number> {
        return parseFloat(await this.scpi.query("SYST:TEMP?"))
    }

    async getReferenceVoltage(): Promise<number> {
        return parseFloat(await this.scpi.query("SYST:VREF?"))
    }

    async beep(frequency: number = 2000, seconds: number = 0.1) {
        await this.scpi.write(`SYST:BEEP ${frequency}Hz, ${seconds.toFixed(3)}s`)
    }

    // Networking commands

    async getIp(): Promise<string> {
        return this.scpi.query("SYST:NET:IPAD?")
    }

    async getMac(): Promise<string> {
        return this.scpi.query("SYST:NET:MAC?")
    }

    async getHostname(): Promise<string> {
        return this.scpi.query("SYST:NET:NAME?")
    }

    async setHostname(name: string) {
        await this.scpi.write(`SYST:NET:NAME "${name}"`)
    }

    async getAddress(): Promise<string|null> {
        const addr = await this.scpi.query("SYST:NET:NAME:RES?")
        return addr.length ? addr : null;
    }

    // Aux serial commands
    
    async openSerial(baud: number = 9600) {
        await this.scpi.write(`AUX:SER:BAUD ${baud}`)
        await this.scpi.write("AUX:SER:ENA ON")
    }

    async closeSerial() {
        await this.scpi.write("AUX:SER:ENA OFF")
    }

    async getSerialRemaining(): Promise<number> {
        return parseInt(await this.scpi.query("AUX:SER:READ?"))
    }

    async writeSerial(payload: Buffer) {
        await this.scpi.write(`AUX:SER:WRITE ${bytes2hex(payload)}`)
    }

    async setDirPin(enable: boolean, activeHigh: boolean = false) {
        var mode = !enable ? "OFF" : ( activeHigh ? "TX" : "NTX" )
        await this.scpi.write(`AUX:SER:DIR ${mode}`)
    }

    private async readSerial(count: number = 256): Promise<Buffer> {
        let payload = await this.scpi.query(`AUX:SER:READ ${count}`)
        return hex2bytes(payload)
    }

    async readSerialLine(delimiter: string = "\n", timeout: number = 0) {
        const start = Date.now();
        while (true) {
            this.serialBuffer = Buffer.concat([this.serialBuffer, await this.readSerial()]);
            const index = this.serialBuffer.indexOf(delimiter);
            if (index >= 0) {
                const line = this.serialBuffer.subarray(0, index).toString();
                this.serialBuffer = this.serialBuffer.subarray(index + Buffer.from(delimiter).length);
                return line;
            }
            if ((Date.now() - start) > timeout) {
                return null;
            }
            await delay(50);
        }
    }

    async readSerialBytes(count: number, timeout: number = 0) {
        const start = Date.now();
        while (true) {
            this.serialBuffer = Buffer.concat([this.serialBuffer, await this.readSerial()]);
            if (this.serialBuffer.length >= count) {
                const data = this.serialBuffer.subarray(0, count);
                this.serialBuffer = this.serialBuffer.subarray(count);
                return data;
            }
            if ((Date.now() - start) > timeout) {
                const data = this.serialBuffer;
                this.serialBuffer = Buffer.alloc(0);
                return data;
            }
            await delay(50);
        }
    }

    // Aux I2C commands

    async openI2c(speed: number = 100000) {
        await this.scpi.write(`AUX:IIC:SPEED ${speed}`)
        await this.scpi.write("AUX:IIC:ENA ON")
    }

    async closeI2c() {
        await this.scpi.write("AUX:IIC:ENA OFF");
    }

    async readI2c(address: number, toRead: number) {
        const result = await this.scpi.query(`AUX:IIC:READ ${address}, ${toRead}`);
        return hex2bytes(result);
    }

    async writeI2c(address: number, payload: Buffer) {
        const result = await this.scpi.query(`AUX:IIC:WRITE ${address}, ${bytes2hex(payload)}`);
        return result === "ON";
    }

    async transferI2c(address: number, payload: Buffer, toRead: number) {
        const result = await this.scpi.query(`AUX:IIC:TRAN ${address}, ${bytes2hex(payload)}, ${toRead}`);
        return hex2bytes(result);
    }

    async scanI2c(address: number) {
        const result = await this.scpi.query(`AUX:IIC:SCAN ${address}`);
        return result === "ON";
    }

    async scanAllI2c() {
        const foundDevices = [];
        for (let addr = 0; addr < 0x80; addr++) {
            if (await this.scanI2c(addr)) {
                foundDevices.push(addr);
            }
        }
        return foundDevices;
    }

    async readEeprom(count: number = 16, address: number = 0x50, offset: number = 0) {
        return await this.transferI2c(address, Buffer.from([offset]), count);
    }

    async writeEeprom(payload: Buffer, address: number = 0x50, offset: number = 0, pagesize: number = 8) {
        let index = 0;
        const end = offset + payload.length;
        while (offset < end) {
            const pageEnd = (offset & ~(pagesize - 1)) + pagesize;
            const chunk = Math.min(pageEnd, end) - offset;
            
            if (!(await this.writeI2c(address, Buffer.concat([Buffer.from([offset]), payload.subarray(index, index + chunk)])))) {
                return false;
            }
            
            offset += chunk;
            index += chunk;
            
            for (let i = 0; i < 10; i++) {
                if (await this.scanI2c(address)) {
                    break;
                }
                await delay(50);
            }
        }
        return true;
    }

    // Aux GPIO commands

    async setGpio(pin: number, state: boolean) {
        assertChannel(pin, 8)
        await this.scpi.write(`AUX:GPIO${pin}:SET ${state?"ON":"OFF"}`)
    }

    async readGpio(pin: number) {
        assertChannel(pin, 8)
        return (await this.scpi.query(`AUX:GPIO${pin}:READ?`)) === "ON"
    }

    // Aux servo commands

    async setServo(pin: number, microseconds: number = 1500) {
        assertChannel(pin, 8)
        await this.scpi.write(`AUX:GPIO${pin}:SERV:SET ON, ${microseconds.toFixed(0)}us`)
    }

    async setServoEnable(pin: number, enable: boolean) {
        assertChannel(pin, 8)
        await this.scpi.write(`AUX:GPIO${pin}:SERV:ENA ${enable?"ON":"OFF"}`)
    }

    async setServoPulse(pin: number, microseconds: number) {
        assertChannel(pin, 8)
        await this.scpi.write(`AUX:GPIO${pin}:SERV:PULS ${microseconds.toFixed(0)}us`)
    }

    async setServoAngle(pin: number, degrees: number) {
        let microseconds = ((degrees / 180.0) * 1000) + 1000
        return this.setServoPulse(pin, microseconds)
    }

    static async findUsbDevices(): Promise<string[]> {
        const results = [];
        const ports = await SerialPort.list();
        for (const port of ports) {
            // WARN: vid/pid are case sensitive
            if (port.vendorId == '16d0' && port.productId == '13fd') {
                results.push(`tty://${port.path}`);
            }
        }
        return results;
    }

}

async function delay(ms: number) {
    return new Promise(resolve => setTimeout(resolve, 50));
}

function assertChannel(channel: number, maximum: number) {
    const valid = Number.isInteger(channel) && channel > 0 && channel <= maximum;
    if (!valid) {
        throw new Error(`Channel must be integer [1..${maximum}]`)
    }
}

function bytes2hex(payload: Buffer): string {
    return payload.reduce((output, elem) => 
        (output + ('0' + elem.toString(16)).slice(-2)),
        '');
}

function hex2bytes(hex: string): Buffer {
    const count = hex.length / 2
    let bytes = Buffer.alloc(count);
    for (let i = 0; i < count; i++) {
        bytes[i] = parseInt(hex.substring(i*2, i*2 + 2), 16);
    }
    return bytes;
}

export default Nautilus

File: python/nautilus/Nautilus.py
from . import SCPI
import time


class Nautilus():
    def __init__(self, uri: str = "tcp://nautilus.local"):
        self.scpi = SCPI.from_uri(uri)
        self.version = 0.0
        self.input_freq: float|None = None

    def __enter__(self):
        self.open()
        return self
    
    def __exit__(self, *args):
        self.close()

    def get_version(self) -> str:
        idn = self.scpi.get_idn()
        # TL Embedded, Nautilus, 001C00484331500120373358, v0.1
        if idn.startswith("TL Embedded, Nautilus,"):
            return float(idn.split(", ")[-1][1:])
        raise Exception("Nautilus not found")

    def open(self):
        self.version = self.get_version()

    def close(self, reset: bool = True):
        if reset:
            self.reset()
        self.scpi.close()

    def reset(self):
        self.scpi.write("*RST")

    # ADC input commands

    def set_input_frequency(self, frequency: float):
        if self.version >= 1.0:
            self.scpi.write(f"INP:FREQ {frequency:.3f}Hz")
        self.input_freq = frequency

    def get_input_voltage(self, channel: int, frequency: float|None = None) -> float:
        if frequency is not None and self.input_freq != frequency:
            self.set_input_frequency(frequency)
        return float(self.scpi.query(f"INP{channel}:VOLT?"))
    
    # DAC output commands

    def set_output_voltage(self, channel: int, volts: float ):
        self.scpi.write( f"OUT{channel}:VOLT {volts:.3f}V" )

    def set_output_enable(self, channel: int, enable: bool):
        self.scpi.write( f"OUT{channel}:ENA {'ON' if enable else 'OFF'}" )

    def set_output(self, channel: int, enable: bool, volts: float = 0.0):
        self.scpi.write( f"OUT{channel}:SET {'ON' if enable else 'OFF'}, {volts:.3f}V" )

    # PSU output commands

    def set_psu_voltage(self, channel: int, volts: float):
        self.scpi.write(f"POW{channel}:VOLT {volts:.3f}V")
        
    def set_psu_current(self, channel: int, amps: float):
        self.scpi.write(f"POW{channel}:CURR {amps:.3f}A")
    
    def set_psu_enable(self, channel: int, enable: bool):
        self.scpi.write(f"POW{channel}:ENA {'ON' if enable else 'OFF'}")
    
    def set_psu(self, channel: int, enable: bool, volts: float, amps: float):
        if self.version >= 1.0:
            self.scpi.write(f"POW{channel}:SET {'ON' if enable else 'OFF'}, {volts:.3f}V, {amps:.3f}A")
        else:
            self.set_psu_voltage(channel, volts)
            self.set_psu_current(channel, amps)
            self.set_psu_enable(channel, enable)
        

    def get_psu_voltage(self, channel: int) -> float:
        return float(self.scpi.query(f"POW{channel}:MEAS:VOLT?"))
    
    def get_psu_current(self, channel: int) -> float:
        return float(self.scpi.query(f"POW{channel}:MEAS:CURR?"))

    # System commands

    def get_vin(self) -> float:
        return float(self.scpi.query("VIN?"))
    
    def get_temperature(self) -> float:
        return float(self.scpi.query("SYST:TEMP?"))
    
    def get_reference_voltage(self) -> float:
        return float(self.scpi.query("SYST:VREF?"))
    
    def beep(self, freq_hz: int = 2000, duration_s: float = 0.1):
        self.scpi.write(f"SYST:BEEP {freq_hz}Hz, {duration_s}s")
    
    # Networking commands
    
    def get_ip(self) -> str:
        return self.scpi.query("SYST:NET:IPAD?")
    
    def get_mac(self) -> str:
        return self.scpi.query("SYST:NET:MAC?")
    
    def get_hostname(self) -> str:
        return self.scpi.query("SYST:NET:NAME?")
    
    def set_hostname(self, name: str):
        self.scpi.write(f"SYST:NET:NAME \"{name}\"")

    def get_address(self) -> str | None:
        address = self.scpi.query("SYST:NET:NAME:RES?")
        return address if address else None

    # Aux serial commands

    def open_serial(self, baud: int = 9600):
        self.scpi.write(f"AUX:SER:BAUD {int(baud)}")
        self.scpi.write("AUX:SER:ENA ON")
        self._serial_buffer = bytearray()

    def close_serial(self):
        self.scpi.write("AUX:SER:ENA OFF")

    def _read_serial(self, count: int = 256) -> bytes:
        result = self.scpi.query(f"AUX:SER:READ {count}")
        return bytes.fromhex(result)
    
    def get_serial_remaining(self) -> int:
        return int(self.scpi.query(f"AUX:SER:READ?"))

    def write_serial(self, payload: bytes):
        self.scpi.write(f"AUX:SER:WRITE {payload.hex()}")

    def set_dir_pin(self, enable: bool, active_high: bool = False):
        if not enable:
            mode = "OFF"
        else:
            mode = "TX" if active_high else "NTX"
        self.scpi.write(f"AUX:SER:DIR {mode}")

    def read_serial_line(self, delimiter: bytes = b"\n", timeout: float = 0.0) -> None | bytes:
        start = time.time()
        while True:
            self._serial_buffer.extend(self._read_serial())
            index = self._serial_buffer.find(delimiter)
            if index >= 0:
                line = bytes(self._serial_buffer[:index])
                self._serial_buffer = self._serial_buffer[index + len(delimiter):]
                return line
            if time.time() - start > timeout:
                return None
            time.sleep(0.05)

    def read_serial_bytes(self, count: int, timeout: float = 0.0) -> bytes:
        start = time.time()
        while True:
            self._serial_buffer.extend(self._read_serial())
            if len(self._serial_buffer) >= count:
                line = bytes(self._serial_buffer[:count])
                self._serial_buffer = self._serial_buffer[count:]
                return line
            if time.time() - start > timeout:
                line = bytes(self._serial_buffer)
                self._serial_buffer = bytearray()
                return line
            time.sleep(0.05)

    # Aux I2C commands

    def open_i2c(self, speed: int = 100000):
        self.scpi.write(f"AUX:IIC:SPEED {int(speed)}")
        self.scpi.write(f"AUX:IIC:ENA ON")

    def close_i2C(self):
        self.scpi.write("AUX:IIC:ENA OFF")

    def read_i2c(self, address: int, to_read: int) -> bytes:
        result = self.scpi.query(f"AUX:IIC:READ {address}, {to_read}")
        return bytes.fromhex(result)
    
    def write_i2c(self, address: int, payload: bytes) -> bool:
        result = self.scpi.query(f"AUX:IIC:WRITE {address}, {payload.hex()}")
        return result == "ON"
    
    def transfer_i2c(self, address: int, payload: bytes, to_read: int) -> bytes:
        result = self.scpi.query(f"AUX:IIC:TRAN {address}, {payload.hex()}, {to_read}")
        return bytes.fromhex(result)
    
    def scan_i2c(self, address: int) -> bool:
        result = self.scpi.query(f"AUX:IIC:SCAN {address}")
        return result == "ON"

    def scan_all_i2c(self) -> list[int]:
        return [ addr for addr in range(0x80) if self.scan_i2c(addr)]
    
    def read_eeprom(self, count: int = 16, address: int = 0x50, offset: int = 0) -> bytes:
        return self.transfer_i2c(address, bytes([offset]), count )
    
    # Default settings support the 24AA01 EEPROMs
    def write_eeprom(self, payload: bytes, address: int = 0x50, offset: int = 0, pagesize: int = 8) -> bool:
        index = 0
        end = offset + len(payload)
        while offset < end:

            page_end = (offset & ~(pagesize-1)) + pagesize
            chunk = min( page_end, end ) - offset

            # Write chunk
            if not self.write_i2c(address, bytes([offset]) + payload[index:index+chunk] ):
                return False
            
            offset += chunk
            index += chunk
            
            # Device acks once page write is complete.
            for _ in range(10):
                if self.scan_i2c(address):
                    break
                time.sleep(0.05)
            
        return True
    
    # Aux GPIO commands

    def set_gpio(self, pin: int, state: bool):
        self.scpi.write(f"AUX:GPIO{pin}:SET {'ON' if state else 'OFF'}")

    def read_gpio(self, pin: int):
        return self.scpi.query(f"AUX:GPIO{pin}:READ?") == 'ON'

    # Aux Servo commands

    def set_servo(self, pin: int, period_us: int = 1500):
        self.scpi.write(f"AUX:GPIO{pin}:SERV:SET ON, {period_us}us")

    def set_servo_enable(self, pin: int, enable: bool):
        self.scpi.write(f"AUX:GPIO{pin}:SERV:ENA {'ON' if enable else 'OFF'}")

    def set_servo_pulse(self, pin: int, period_us: int):
        self.scpi.write(f"AUX:GPIO{pin}:SERV:PULS {period_us}us")

    def set_servo_angle(self, pin: int, angle_deg: float):
        period_us = ((angle_deg / 180.0) * 1000) + 1000
        self.set_servo_pulse(pin, period_us)

    # Calibration commands
    # These should only be used with the appropriate jig...
    
    def calibrate_output(self, channel: int, setpoint_a: float = 1.0, setpoint_b: float = 11.0):
        self.set_output_enable(channel, True)
        self.scpi.write( f"OUT{channel}:VOLT:UNCAL {setpoint_a:.3f}V" )
        time.sleep(0.5)
        result_a = self.get_input_voltage(channel)
        self.scpi.write( f"OUT{channel}:VOLT:UNCAL {setpoint_b:.3f}V" )
        time.sleep(0.5)
        result_b = self.get_input_voltage(channel)
        self.set_output_enable(channel, False)
        self.scpi.write( f"OUT{channel}:VOLT:CAL {setpoint_a:.3f}V, {result_a:.3f}V, {setpoint_b:.3f}V, {result_b:.3f}V" )

    @staticmethod
    def find_usb_devices() -> list[str]:
        results = []
        from serial.tools.list_ports import comports
        for port in comports():
            if port.vid == 0x16D0 and port.pid == 0x13FD:
                results.append(f"tty://{port.device}")
        return results
